### PR TITLE
Add pod handling APIs

### PR DESCRIPTION
**Pull request by Rohit-PX**

### CODE REVIEW

File: drivers/scheduler/k8s/k8s.go [hunk 1]
```diff
@@ -540,6 +561,25 @@ func (k *k8s) destroyCoreObject(spec interface{}, opts map[string]bool, app *spe
 		}
 
 		logrus.Infof("[%v] Destroyed Service: %v", app.Key, obj.Name)
+	} else if obj, ok := spec.(*v1.Pod); ok {
+		if value, ok := opts[scheduler.OptionsWaitForResourceLeakCleanup]; ok && value {
+			logrus.Infof("RK => VALUE: %v", value)
```
[the remark this answers is not in the document]
Comment (Rohit-PX) [reply drivers/scheduler/k8s/k8s.go]: Fixed.

File: drivers/scheduler/k8s/k8s.go [hunk 2]
```diff
@@ -540,6 +561,25 @@ func (k *k8s) destroyCoreObject(spec interface{}, opts map[string]bool, app *spe
...
+			logrus.Infof("RK => OK: %v", ok)
+			if podList, err = k8sOps.GetPods(obj.Namespace); err != nil {
+				logrus.Warnf("[%v] Error getting pods. Err: %v", app.Key, err)
+			}
+		}
+		logrus.Infof("POD LIST: %v", podList)
```
[the remark this answers is not in the document]
Comment (Rohit-PX) [reply drivers/scheduler/k8s/k8s.go]: Fixed.

File: drivers/scheduler/k8s/k8s.go [hunk 3]
```diff
@@ -490,6 +492,24 @@ func (k *k8s) createCoreObject(spec interface{}, ns *v1.Namespace, app *spec.App
 
 		logrus.Infof("[%v] Created Secret: %v", app.Key, secret.Name)
 		return secret, nil
+	} else if obj, ok := spec.(*v1.Pod); ok {
+		obj.Namespace = ns.Name
+		pod, err := k8sOps.CreatePod(obj)
+		if errors.IsAlreadyExists(err) {
+			if pods, err := k8sOps.GetPods(obj.Namespace); err == nil {
+				logrus.Infof("[%v] Found existing Pods: %v", app.Key, pods)
+				return pod, nil
```
[the remark this answers is not in the document]
Comment (Rohit-PX) [reply drivers/scheduler/k8s/k8s.go]: Using GetPodByName to get that specific pod.

File: drivers/scheduler/k8s/specs/wordpress/installer.yaml [hunk 4]
```diff
@@ -1,54 +1,34 @@
...
+      valueFrom:
+         secretKeyRef:
+           name: mysql-pass
+           key: password
+    volumeMounts:
+    - name: wordpress-persistent-storage
+      mountPath: /wordpress
+  volumes:
+  - name: wordpress-persistent-storage
+    persistentVolumeClaim:
```
[the remark this answers is not in the document]
Comment (Rohit-PX) [reply drivers/scheduler/k8s/specs/wordpress/installer.yaml]: Done.

File: drivers/scheduler/k8s/k8s.go [hunk 5]
```diff
@@ -571,6 +591,21 @@ func (k *k8s) destroyCoreObject(spec interface{}, opts map[string]bool, app *spe
 		}
 
 		logrus.Infof("[%v] Destroyed StorkRule: %v", app.Key, obj.Name)
+	} else if obj, ok := spec.(*v1.Pod); ok {
+		if value, ok := opts[scheduler.OptionsWaitForResourceLeakCleanup]; ok && value {
+			if pods, err = k8sOps.GetPods(obj.Namespace); err != nil {
```
Comment (harsh-px) [inline drivers/scheduler/k8s/k8s.go]: this should do a get pod on the pod being deleted and append to the pods list. We use that pod list below to wait for cleanup

File: drivers/scheduler/k8s/k8s.go [hunk 6]
```diff
@@ -630,6 +665,16 @@ func (k *k8s) WaitForRunning(ctx *scheduler.Context) error {
 			}
 
 			logrus.Infof("[%v] Validated StorkRule: %v", ctx.App.Key, svc.Name)
+		} else if obj, ok := spec.(*v1.Pod); ok {
+			isPodReady := k8sOps.IsPodReady(*obj)
```
Comment (harsh-px) [inline drivers/scheduler/k8s/k8s.go]: nit: change to `if ! k8sOps.IsPodReady(*obj) {`

File: drivers/scheduler/k8s/k8s.go [hunk 7]
```diff
@@ -856,6 +910,19 @@ func (k *k8s) GetVolumeParameters(ctx *scheduler.Context) (map[string]map[string
 			result[snapData.Spec.VolumeSnapshotDataSource.PortworxSnapshot.SnapshotID] = map[string]string{
 				SnapshotParent: snap.Spec.PersistentVolumeClaimName,
 			}
+		} else if obj, ok := spec.(*snap_v1.VolumeSnapshot); ok {
```
Comment (harsh-px) [inline drivers/scheduler/k8s/k8s.go]: This PR should not have any snapshot changes. Snapshot support is already present in master

File: vendor/github.com/portworx/sched-ops/k8s/k8s.go [hunk 8]
```diff
@@ -91,9 +91,9 @@ type NamespaceOps interface {
 	// GetNamespace returns a namespace object for given name
 	GetNamespace(name string) (*v1.Namespace, error)
 	// CreateNamespace creates a namespace with given name and metadata
-	CreateNamespace(name string, metadata map[string]string) (*v1.Namespace, error)
+	CreateNamespace(string, map[string]string) (*v1.Namespace, error)
```
Comment (harsh-px) [inline vendor/github.com/portworx/sched-ops/k8s/k8s.go]: The PR has vendor changes with changes to glide.lock

Make sure you update torpedo vendor using `make vendor` rule and then update this  PR

File: drivers/scheduler/k8s/k8s.go (hunk 7 above)
Comment (Rohit-PX) [reply drivers/scheduler/k8s/k8s.go]: Removed these lines code.

File: drivers/scheduler/k8s/k8s.go (hunk 6 above)
Comment (Rohit-PX) [reply drivers/scheduler/k8s/k8s.go]: Fixed.

File: drivers/scheduler/k8s/k8s.go (hunk 5 above)
Comment (Rohit-PX) [reply drivers/scheduler/k8s/k8s.go]: Using GetPodByName to get that pod and adding it to the list of pods and then assigning that list to the interface 
'pods' that will be returned.

File: drivers/scheduler/k8s/specs/wordpress/installer.yaml [hunk 9]
```diff
@@ -1,54 +1,34 @@
...
-          initialDelaySeconds: 5
-          periodSeconds: 5
-      volumes:
-      - name: wordpress-persistent-storage
-        persistentVolumeClaim:
-          claimName: wp-pv-claim
+  securityContext:
+    fsGroup: 1000
+  containers:
+  - image: samuelmyers/wp-cli:1.2.25
```
[the remark this answers is not in the document]
Comment (Rohit-PX) [reply drivers/scheduler/k8s/specs/wordpress/installer.yaml]: Updated to => samuelmyers/wp-cli:1.2.28

File: drivers/scheduler/k8s/specs/wordpress/installer.yaml [hunk 10]
```diff
@@ -1,54 +1,34 @@
...
+    - name: WORDPRESS_DB_PASSWORD
+      valueFrom:
+         secretKeyRef:
+           name: mysql-pass
+           key: password
+    volumeMounts:
+    - name: wordpress-persistent-storage
+      mountPath: /wordpress
+  volumes:
+  - name: wordpress-persistent-storage
```
[the remark this answers is not in the document]
Comment (Rohit-PX) [reply drivers/scheduler/k8s/specs/wordpress/installer.yaml]: Added readiness Probe.

File: drivers/scheduler/k8s/specs/wordpress/load-driver.yaml [hunk 11]
```diff
@@ -1,36 +1,23 @@
 apiVersion: apps/v1beta2
```
[the remark this answers is not in the document]
Comment (Rohit-PX) [reply drivers/scheduler/k8s/specs/wordpress/load-driver.yaml]: Fixed.

File: drivers/scheduler/k8s/k8s.go [hunk 12]
```diff
@@ -571,6 +592,24 @@ func (k *k8s) destroyCoreObject(spec interface{}, opts map[string]bool, app *spe
 		}
 
 		logrus.Infof("[%v] Destroyed StorkRule: %v", app.Key, obj.Name)
+	} else if obj, ok := spec.(*v1.Pod); ok {
+		if value, ok := opts[scheduler.OptionsWaitForResourceLeakCleanup]; ok && value {
+			pod, err := k8sOps.GetPodByName(obj.Name, obj.Namespace)
+			if err != nil {
+				logrus.Warnf("[%v] Error getting pods. Err: %v", app.Key, err)
+			}
+			podList = append(podList, pod)
```
Comment (harsh-px) [inline drivers/scheduler/k8s/k8s.go]: why not directly append to pods here?

File: drivers/scheduler/k8s/specs/wordpress/installer.yaml (hunk 9 above)
[the remark this answers is not in the document]
Comment (harsh-px) [reply drivers/scheduler/k8s/specs/wordpress/installer.yaml]: I still see wp-cli:1.2.25 here

File: drivers/scheduler/k8s/k8s.go (hunk 12 above)
Comment (Rohit-PX) [reply drivers/scheduler/k8s/k8s.go]: We get the following error in that case:
pods = append(pods, pod)
"first argument to append must be slice; have interface {}"

File: drivers/scheduler/k8s/specs/wordpress/installer.yaml (hunk 9 above)
[the remark this answers is not in the document]
Comment (Rohit-PX) [reply drivers/scheduler/k8s/specs/wordpress/installer.yaml]: Fixed.

**Pull request merged**